streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -1848,113 +1907,101 @@ object ZChannel {
-                           } yield !errored
+                           pull.flatMap(channel =>
+                             for {
+                               latch <- Promise.make[Nothing, Unit]
+                               raceIOs =
+                                 ZIO.scopedWith { scope =>
+                                   (incoming >>> channel)
+                                     .toPullInAlt(scope)
+                                     .flatMap(evaluatePull)
+                                 }
+                               _ <- permits
+                                      .withPermit(latch.succeed(()) *> raceIOs)
+                                      .fork
+                               _ <- latch.await
+                             } yield ()
+                           )
                          case MergeStrategy.BufferSliding =>
-                           for {
-                             canceler <- Promise.make[Nothing, Unit]
-                             latch    <- Promise.make[Nothing, Unit]
-                             size     <- cancelers.size
-                             _        <- ZIO.when(size >= n)(cancelers.take.flatMap(_.succeed(())))
-                             _        <- cancelers.offer(canceler)
-                             raceIOs =
-                               ZIO.scopedWith { scope =>
-                                 (queueReader >>> channel)
-                                   .toPullIn(scope)
-                                   .flatMap(
-                                     evaluatePull(_)
-                                       .raceAwait(errorSignal.await.interruptible)
-                                       .raceAwait(canceler.await.interruptible)
-                                   )
-                               }
-                             childFiber <- permits
-                                             .withPermit(latch.succeed(()) *> raceIOs)
-                                             .forkIn(scope)
-                             _       <- latch.await
-                             errored <- errorSignal.isDone
-                           } yield !errored
+                           pull.flatMap(channel =>
+                             for {
+                               canceler <- Promise.make[Nothing, Unit]
+                               latch    <- Promise.make[Nothing, Unit]
+                               size     <- cancelers.size
+                               _        <- ZIO.when(size >= n)(cancelers.take.flatMap(_.succeed(())))
+                               _        <- cancelers.offer(canceler)
+                               raceIOs =
+                                 ZIO.scopedWith { scope =>
+                                   (incoming >>> channel)
+                                     .toPullInAlt(scope)
+                                     .flatMap(evaluatePull(_).race(canceler.await))
```

Comment:
Don't we need to make the canceler interruptible here?

Reply:
I don't see why, it's just a promise

Reply:
The issue is that `race` will attempt to interrupt the loser and awaiting for its completion. If we're draining the `ZChannel` within an uninterruptible region, then this might hang waiting for the interruption of the `Promise#await` which will never happen.

PS: I just realised that the issue is mostly with `errorSignal.await`, but I think we should also the canceller interruptible. Simple reproducers where the test currently timeouts due to `errorSignal.await`:

```scala
      test("mapOutZIOPar in uninterruptible region") {
        for {
          _ <- ZChannel.unit.mapOutZIOPar(4)(_ => ZIO.unit).runDrain.uninterruptible
        } yield assertCompletes
      } @@ timeout(5.seconds),
      test("mergeAllWith in uninterruptible region") {
        for {
          _ <- ZChannel.mergeAllWith(ZChannel.unit, 4, mergeStrategy = MergeStrategy.BufferSliding)((l, _) => l).runDrain.uninterruptible
        } yield assertCompletes
      } @@ timeout(5.seconds),
```

Tests pass when we change to `errorSignal.await.interruptible` (even without changing the canceller), but I think we should probably also make the canceller interruptible as there might be other usecases that will hang with the current code

Reply:
I don't think anybody uses `BufferSliding` because the implementation is clearly wrong, I think it should be deprecated. At the very least, I can replace the queue with a sorted map to provide the correct semantics.